sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoApiClient.kt

```diff
@@ -0,0 +1,180 @@
+        )
+    }
+
+    /**
+     * Awaken queued requests from persistent store
+     */
+    fun restoreQueue() {
+        // TODO JSON type safety checks
+        Registry.dataStore.fetch(QUEUE_KEY)?.let {
+            val queue = JSONArray(it)
+            Array(queue.length()) { i -> queue.optString(i) }
+        }?.map { uuid ->
+            Registry.dataStore.fetch(uuid)?.let { json ->
+                val jsonObject = JSONObject(json)
+                val request = KlaviyoApiRequest.fromJson(jsonObject)
+                enqueueRequest(request)
+            }
+        }
+    }
+
+    /**
+     * Initialize the network queue batching with a thread to run on
+     */
+    private fun initBatch() {
+        if (!handlerThread.isAlive) {
+            handlerThread.start()
+            handler = HandlerUtil.getHandler(handlerThread.looper)
+        }
+
+        handler?.post(NetworkRunnable())
+    }
+
+    /**
+     * Gets the current size of the API queue
+     *
+     * @return number of requests in the batch queue
+     */
+    fun getQueueSize(): Int = apiQueue.size
+
+    /**
+     * Force the API queue to send immediately instead of waiting on the configured criteria
+     */
+    private fun flushQueue() {
+        handler?.removeCallbacksAndMessages(null)
+        handler?.post(NetworkRunnable(true))
+    }
+
+    /**
+     * Runnable which flushes the API queue in batches for efficiency.
+     * As long as there are items in the queue, the thread will loop.
+     * When the queue is cleared the thread will not loop itself and will terminate.
+     *
+     * @property force Boolean that will force the queue to flush now
+     */
+    class NetworkRunnable(val force: Boolean = false) : Runnable {
+        private val queueInitTime = Registry.clock.currentTimeMillis()
+
+        private val flushInterval: Long
+            get() = Registry.config.networkFlushInterval.toLong()
+
+        private val flushDepth: Int
+            get() = Registry.config.networkFlushDepth
+
+        override fun run() {
+            val emptied = flushQueue(force)
+            if (!emptied) {
+                handler?.postDelayed(this, flushInterval)
+            }
+        }
+
+        /**
+         * Empties all queued requests by instantly sending each request to Klaviyo
+         * The queue will flush whenever the triggers specified in config are met
+         *
+         * @param force Overrides the typical timings to force the queue to flush
+         *
+         * @return Whether the request queue was emptied or not
+         */
+        private fun flushQueue(force: Boolean = false): Boolean {
+            val queueTimePassed = Registry.clock.currentTimeMillis() - queueInitTime
+
+            if (force || getQueueSize() >= flushDepth || queueTimePassed >= flushInterval) {
+                while (!apiQueue.isEmpty()) {
+                    val apiRequest = apiQueue.poll()
+                    apiRequest?.apply {
+                        send()
```

Comment:
Do you need to handle the return value here? Seems like sometimes you return null like when the network is disconnected. Maybe this is for a future iteration?

Reply:
ooh shoot yeah good call this will clear the API queue even if the requests don't send!

Reply:
This is the other reason I'd re-opened the network layer ticket BTW, because I'm totally missing things like retrying rate limit errors or handling other error types. But I didn't even notice that I'm straight up throwing requests while offline 🤦

Reply:
I had realizations like when I was working on the iOS SDK. Don't beat yourself up there's a lot to think about. :)

Reply:
Thanks for the reassurance, its definitely more complicated than it seems from the surface!

Reply:
Resolving, see PR #33